File: data-structures/queue/implementation.cpp
#include <iostream>
using namespace std;

class Queue {
private:
    int* arr;
    int frontIndex;
    int rearIndex;
    int capacity;
    int size;

public:
    Queue(int cap = 100) {
        capacity = cap;
        arr = new int[capacity];
        frontIndex = 0;
        rearIndex = -1;
        size = 0;
    }

    ~Queue() {
        delete[] arr;
    }

    void enqueue(int val) {
        if (size == capacity) return;

        rearIndex = (rearIndex + 1) % capacity;
        arr[rearIndex] = val;
        size++;
    }

    void dequeue() {
        if (size == 0) return;

        frontIndex = (frontIndex + 1) % capacity;
        size--;
    }

    int front() {
        if (size == 0) return -1;
        return arr[frontIndex];
    }

    bool isEmpty() {
        return size == 0;
    }
};

File: data-structures/queue/implementation.java
public class Queue {
    private int[] arr;
    private int front;
    private int rear;
    private int size;
    private int capacity;

    public Queue(int cap) {
        capacity = cap;
        arr = new int[capacity];
        front = 0;
        rear = -1;
        size = 0;
    }

    public void enqueue(int val) {
        if (size == capacity) return;

        rear = (rear + 1) % capacity;
        arr[rear] = val;
        size++;
    }

    public void dequeue() {
        if (size == 0) return;

        front = (front + 1) % capacity;
        size--;
    }

    public int front() {
        if (size == 0) return -1;
        return arr[front];
    }

    public boolean isEmpty() {
        return size == 0;
    }
}
